book/bayesian_optimization.qmd

```diff
@@ -0,0 +1,844 @@
+In the following, we will use `r ref("ObjectiveRFunDt")`:
+
+```{r}
+sinus_1D = function(xdt) {
+  y = 2 * xdt$x * sin(14 * xdt$x)
+  data.table(y = y)
+}
+```
+
+An `r ref("Objective")` always requires the specification of the domain and codomain in the form of a `r ref("ParamSet")`:
+
+```{r}
+domain = ps(x = p_dbl(lower = 0, upper = 1))
+codomain = ps(y = p_dbl(tags = "minimize"))
+objective = ObjectiveRFunDt$new(sinus_1D,
+  domain = domain, codomain = codomain)
+```
+
+We can proceed to visualize the sinusoidal function by generating a grid of points on which we evaluate the function:
+
+```{r}
+xydt = generate_design_grid(domain, resolution = 1001)$data
+xydt[, y := objective$eval_dt(xydt)$y]
+```
+
+```{r}
+#| echo: true
+#| label: fig-bayesian-optimization-sinusoidal
+#| fig-cap: Visualization of the sinusoidal function.
+#| fig-alt: FIXME
+library(ggplot2)
+ggplot(aes(x = x, y = y), data = xydt) +
+  geom_line() +
+  theme_minimal()
+```
+
+The global minimum is located at `x = 0.7918238`:
+
+```{r}
+sinus_1D(data.table(x = 0.7918238))
+```
+
+By wrapping the objective function in an `r ref("OptimInstanceSingleCrit")` we can proceed to optimize it, e.g., via a simple random search.
+
+```{r, output=FALSE}
+instance = OptimInstanceSingleCrit$new(objective,
+  terminator = trm("evals", n_evals = 20))
+optimizer = opt("random_search", batch_size = 20)
+optimizer$optimize(instance)
+```
+
+```{r}
+instance$archive$best()
+```
+
+To see all available optimizers that can be used for black box optimization, you can inspect the following dictionary:
+
+```{r}
+as.data.table(mlr_optimizers)
+```
+
+Note that evolutionary strategies are available within the mlr3 ecosystem via the `miesmuschel` package which we will not cover in this chapter.
+
+## Building Blocks of Bayesian Optimization {#sec-bayesian-optimization-blocks}
+
+Bayesian optimization (BO) is an iterative optimization algorithm that makes use of a so-called surrogate to model the unknown black box function.
+After having observed an initial design of observations, the surrogate model is trained on all data points observed so far and a so-called acquisition function is used to determine which point of the search space is a promising candidate that should be evaluated next.
+The acquisition function relies on the predictions of the surrogate model and requires no evaluation of the true black box function and therefore is comparably cheap to optimize.
+The acquisition function should balance exploration and exploitation of the BO algorithm, i.e., we want to exploit knowledge about regions where we observed that performance is good and the surrogate model has low uncertainty but also want to make sure that we do not miss crucial regions and therefore also want to explore into regions where the uncertainty of the surrogate model is high.
+After having evaluated the next candidate, the process repeats itself until a given termination criteria is met.
```

Comment:
Sorry but this is too complicated to follow. Either break it up with examples, use some maths, or clearly define terms upfront before discussing how they work